cmake/detect_gpu.cmake

```diff
@@ -0,0 +1,221 @@
+
+
+################################################################################
+# A function for automatic detection of GPUs installed 
+# (if autodetection is enabled)
+# Usage:
+#     check_available(arch_var,nvcc_var)
+function(check_available arch_ver nvcc_ver)
+    if (${arch_ver} LESS 30)
+        message(FATAL_ERROR "Do not support gpu architecture < 30")
+    elseif ((${arch_ver} GREATER_EQUAL 70) AND (${nvcc_ver} LESS 9))
+        message(FATAL_ERROR "Volta support: Needs cuda version >= 9.0")
+    endif()
+endfunction()
+
+################################################################################
+# Function for selecting GPU arch flags for nvcc based on CUDA_ARCH_NAME
+# Usage:
+#     ginkgo_select_nvcc_arch_flags(out_variable)
+function(ginkgo_select_nvcc_arch_flags out_variable)
+    set(__nvcc_ver 8) # default cuda 8
+    ginkgo_detect_nvcc_version(__nvcc_ver)
+
+    # List of arch names
+    if (__nvcc_ver GREATER_EQUAL 9)
+        set(__archs_names "Kepler" "Maxwell" "Pascal" "Volta" "All" "Manual")
+        list(APPEND known_gpu_archs "70")
+    else()
+        set(__archs_names "Kepler" "Maxwell" "Pascal" "All" "Manual")
+    endif()
+
+    set(__archs_name_default "Auto")
+    if(NOT CMAKE_CROSSCOMPILING)
+        list(APPEND __archs_names "Auto")
+        set(__archs_name_default "Auto")
+    endif()
+
+    # set CUDA_ARCH_NAME strings (so it will be seen as dropbox in CMake-Gui)
+    set(CUDA_ARCH_NAME ${__archs_name_default} CACHE STRING
+        "Select target NVIDIA GPU achitecture.")
+    set_property( CACHE CUDA_ARCH_NAME PROPERTY STRINGS "" ${__archs_names} )
+
+    # verify CUDA_ARCH_NAME value
+    if(NOT ";${__archs_names};" MATCHES ";${CUDA_ARCH_NAME};")
+        string(REPLACE ";" ", " __archs_names "${__archs_names}")
+        message(FATAL_ERROR "Only ${__archs_names} architeture names
+                            are supported.")
+    endif()
+
+    if(${CUDA_ARCH_NAME} STREQUAL "Manual")
+        set(CUDA_ARCH_BIN ${ginkgo_known_gpu_archs} CACHE STRING
+            "Specify 'real' GPU architectures to build binaries for,
+            BIN(PTX) format is supported")
+        set(CUDA_ARCH_PTX "30" CACHE STRING
+            "Specify 'virtual' PTX architectures
+            to build PTX intermediate code for")
+        mark_as_advanced(CUDA_ARCH_BIN CUDA_ARCH_PTX)
+    else()
+        unset(CUDA_ARCH_BIN CACHE)
+        unset(CUDA_ARCH_PTX CACHE)
+    endif()
+
+    if(${CUDA_ARCH_NAME} STREQUAL "Kepler")
+        set(__cuda_arch_bin "30 35 37")
+    elseif(${CUDA_ARCH_NAME} STREQUAL "Maxwell")
+        set(__cuda_arch_bin "50 52")
+    elseif(${CUDA_ARCH_NAME} STREQUAL "Pascal")
+        set(__cuda_arch_bin "60 61")
+    elseif(${CUDA_ARCH_NAME} STREQUAL "Volta")
+        set(__cuda_arch_bin "70")
+    elseif(${CUDA_ARCH_NAME} STREQUAL "All")
+        set(__cuda_arch_bin ${ginkgo_known_gpu_archs})
+    elseif(${CUDA_ARCH_NAME} STREQUAL "Auto")
+        ginkgo_detect_installed_gpus(__cuda_arch_bin)
```

Comment:
A more elegant solution, without the need for if's is to build an associative array:

```cmake
set(_cuda_arch_bin_Kepler ...)
set(_cuda_arch_bin_Maxwell ...)
set(_cuda_arch_bin_Pascal ...)
...
```
And then just do: `set(__cuda_arch_bin ${__cuda_arch_bin_${CUDA_ARCH_NAME}})`.